server/server.go

```diff
@@ -514,6 +516,82 @@ func (s *server) recvLoop(stream rpcpb.ControlService_StreamStatusServer) error
 	}
 }
 
+func (s *server) AddNode(ctx context.Context, req *rpcpb.AddNodeRequest) (*rpcpb.AddNodeResponse, error) {
+	zap.L().Debug("received add node request", zap.String("name", req.Name))
+
+	if info := s.getClusterInfo(); info == nil {
+		return nil, ErrNotBootstrapped
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var logLevel, whitelistedSubnets, pluginDir string
+
+	nodeName := req.Name
+	if _, exists := s.network.nodeInfos[nodeName]; exists {
+		return nil, fmt.Errorf("node with name %s already exists", nodeName)
+	}
+	// user can override bin path for this node...
+	execPath := req.StartRequest.ExecPath
+	if execPath == "" {
+		// ...or use the same binary as the rest of the network
+		execPath = s.network.binPath
+	}
+	_, err := os.Stat(execPath)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return nil, utils.ErrNotExists
+		}
+		return nil, fmt.Errorf("failed to stat exec %q (%w)", execPath, err)
+	}
+	if req.StartRequest.LogLevel != nil {
+		logLevel = *req.StartRequest.LogLevel
+	}
+	if req.StartRequest.WhitelistedSubnets != nil {
+		whitelistedSubnets = *req.StartRequest.WhitelistedSubnets
+	}
+	if req.StartRequest.PluginDir != nil {
+		pluginDir = *req.StartRequest.PluginDir
+	}
+
+	rootDataDir := s.clusterInfo.RootDataDir
+
+	logDir := filepath.Join(rootDataDir, nodeName, "log")
+	dbDir := filepath.Join(rootDataDir, nodeName, "db-dir")
+
+	configFile := createConfigFileString(logLevel, logDir, dbDir, pluginDir, whitelistedSubnets)
+	stakingCert, stakingKey, err := staking.NewCertAndKeyBytes()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't generate staking Cert/Key: %w", err)
+	}
+	nodeConfig := node.Config{
+		Name:               req.Name,
+		ConfigFile:         configFile,
+		ImplSpecificConfig: json.RawMessage(fmt.Sprintf(`{"binaryPath":"%s","redirectStdout":true,"redirectStderr":true}`, execPath)),
+		StakingKey:         string(stakingKey),
+		StakingCert:        string(stakingCert),
+	}
+	s.network.nodeNames = append(s.network.nodeNames, req.Name)
```

Comment:
I think we should do this after we've called `s.network.nw.AddNode` in case that errors